### PR TITLE
Support nullary private method (toggle visibility)

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This is another way to mark methods private that we should support (some
codebases use this style heavily).

Supersedes #3758.

### Test plan

See included automated tests.

### CODE REVIEW

File: namer/namer.cc [hunk 1]
```diff
@@ -490,6 +515,25 @@ class SymbolFinder {
...
+            }
+
+            auto recv = ast::cast_tree<ast::ConstantLit>(original.recv);
+            if (recv == nullptr || recv->symbol != core::Symbols::Sorbet_Private_Static()) {
+                return tree;
+            }
+
+            auto &current = methodVisiStack.back();
+            auto methodName = unwrapLiteralToMethodName(ctx, original.args[1]);
+            foundDefs->addModifier(Modifier{current->kind, current->owner, current->loc, current->name, methodName});
```
Comment (elliottt) [inline namer/namer.cc:536]: What do you think about adding a method called `withTarget` to `Modifier` that copies everything, but overwrites the target parameter? Then this line would become:
```c++
foundDefs->addModifier(current.withTarget(methodName));
```
As it is now, you have to read through all the args to know that it's copying everything except for the target.

File: test/testdata/infer/private_class_methods.rb [hunk 2]
```diff
@@ -44,7 +44,5 @@ def method_i; end
 Test.method_g # error: Non-private call to private method `Test::ClassMethods#method_g`
 Test.method_h # error: Non-private call to private method `Test::ClassMethods#method_h`
 
-# TODO: The following methods should contain errors. Sorbet currently does not support setting method
-# visibility using the private/protected keywords that affect the visibility of subsequent methods.
-Test.method_f
-Test.method_i
+Test.method_f # error: Non-private call to private method `Test.method_f`
+Test.method_i # error: Non-private call to private method `Test::ClassMethods#method_i`
```
Comment (elliottt) [inline test/testdata/infer/private_class_methods.rb:48]: Nice! 🎉

**Review by elliottt — APPROVED** [review]
Some c++ comments, but otherwise this looks great!

File: namer/namer.cc [hunk 3]
```diff
@@ -490,6 +515,25 @@ class SymbolFinder {
                 default:
                     return tree;
             }
+        } else if (original.args.size() == 2) {
+            ENFORCE(!methodVisiStack.empty());
```
[the remark this answers is not in the document]
Comment (jez) [reply namer/namer.cc:519]: it won't—so let's ENFORCE it, yeah?

File: namer/namer.cc [hunk 4]
```diff
@@ -475,7 +484,23 @@ class SymbolFinder {
...
+                case core::Names::public_().rawId():
+                    ENFORCE(!methodVisiStack.empty());
+                    methodVisiStack.pop_back();
+                    methodVisiStack.emplace_back(optional<Modifier>{Modifier{
+                        Modifier::Kind::Method,
+                        getOwner(),
+                        original.loc,
+                        original.fun,
+                        core::NameRef::noName(),
+                    }});
```
[the remark this answers is not in the document]
Comment (jez) [reply namer/namer.cc:500]: i find adding constructors to be a lot of boilerplate that needs to be kept in sync when adding a new field. how strongly do you feel about this?

File: namer/namer.cc [hunk 5]
```diff
@@ -475,7 +484,23 @@ class SymbolFinder {
...
         auto &original = ast::cast_tree_nonnull<ast::Send>(tree);
 
-        if (original.args.size() == 1) {
+        if (original.args.size() == 0) {
+            switch (original.fun.rawId()) {
+                case core::Names::private_().rawId():
+                case core::Names::protected_().rawId():
+                case core::Names::public_().rawId():
+                    ENFORCE(!methodVisiStack.empty());
+                    methodVisiStack.pop_back();
```
[the remark this answers is not in the document]
Comment (jez) [reply namer/namer.cc:493]: sure i can try it

File: namer/namer.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (elliottt) [reply namer/namer.cc:500]: It's a non-blocking request, just a style suggestion.

**Pull request merged**